Implement direct returns

zqianem opened the pull request:
Rough initial implementation that doesn't check nil nor the number of return values, and adds a hacky `"subseq"` to the Python writer.

Looking for ways to avoid some of the hackiness and the proper place to add a check number of return values — exceptions thrown in `to_python.py` don't fulfill `success = false` in the test files.

Closes #5

zqianem committed:
Implement direct returns

Rough initial implementation that doesn't check nil nor the number of
return values, and adds a hacky "subseq" to the Python writer

ktvng commented:
I would suggest not handling this issue in the `to_python` visitor, which occurs at the end of the parse/compile/typecheck workflow. As return statements are currently treated as no-op in the typechecker as well as in other visitors, any code statement of the return -- `return /* statement /*` won't be checked. As you've noted, exceptions thrown in the `to_python` visitor do not yield compiler exceptions. This is by design, as all exceptions are expected to be caught and handled before this step. This step is intended purely as translation.

Instead, I would suggest performing this transformation directly after parsing the source code into an AST. This allows all downstream visitors to perform validations as they do today with no needed changes. Effectively, this transformation would turn 

```
return A, B
```

into

```
name_of_first_return_value = A
name_of_second_return_value = B
return

```

This resulting code is what is expected by all type checking visitors today, so all existing checks would be performed.

zqianem committed:
Implement direct returns correctly

Now also includes check for number of return values

zqianem commented:
Changed the implementation to use a new workflow step and added a proper check for number of return values.

@ktvng please re-review when you get a chance.

ktvng commented on src/eisen/common/exceptions.py:
```diff
@@ -102,3 +102,7 @@ class IncompatibleBinding(AbstractException):
+
+    class ReturnValueCountMismatch(AbstractException):
+        type = "ReturnValueCountMismatch"
+        description = "wrong number of return values provied by direct return"
```
nit: sp

zqianem committed:
Use a custom state instead of misusing context

zqianem committed:
Fix passing of rets field to derived state

zqianem committed:
Refactor direct returns visitor

Separates the looping / child replacement and the actual conversion of
direct returns into multiple statements

zqianem committed:
Fix typo
